### PR TITLE
Fixed Base64 EncodeToUtf8 / DecodeFromUtf8 return states (Done | NeedMoreData)

### PR DESCRIPTION
**Pull request by gfoidl**
Migrated from https://github.com/dotnet/corefx/pull/42395

---

Fixes https://github.com/dotnet/corefx/issues/42245
See also https://github.com/dotnet/corefx/issues/42236

Adapted test according the table in https://github.com/dotnet/corefx/issues/42245#issuecomment-548359376 and fixed the implementation.

### CODE REVIEW

**Comment by stephentoub** [comment]
@GrabYourPitchforks, can you review this?

File: src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs [hunk 1]
```diff
@@ -17,21 +17,23 @@ namespace System.Buffers.Text
...
         /// <param name="bytesWritten">The number of bytes written into the output span. This can be used to slice the output for subsequent calls, if necessary.</param>
-        /// <param name="isFinalBlock">True (default) when the input span contains the entire data to decode.
-        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <param name="isFinalBlock"><see langword="true"/> (default) when the input span contains the entire data to encode.
+        /// Set to <see langword="true"/> when the source buffer contains the entirety of the data to encode.
+        /// Set to <see langword="false"/> if this method is being called in a loop and if more input data may follow.
+        /// At the end of the loop, call this (potentially with an empty source buffer) passing <see langword="true"/>.</param>
         /// <returns>It returns the OperationStatus enum values:
-        /// - Done - on successful processing of the entire input span
+        /// - Done - on successful processing of the entire input span (also when <paramref name="isFinalBlock"/> is set to <see langword="false"/>)
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs]: IMO the parenthetical is unnecessary and possibly confusing. The docs for _NeedMoreData_ and _InvalidData_ already call out these cases.

File: src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs [hunk 2]
```diff
@@ -19,19 +19,21 @@ namespace System.Buffers.Text
...
         /// <param name="bytesWritten">The number of bytes written into the output span. This can be used to slice the output for subsequent calls, if necessary.</param>
-        /// <param name="isFinalBlock">True (default) when the input span contains the entire data to encode.
-        /// Set to false only if it is known that the input span contains partial data with more data to follow.</param>
+        /// <param name="isFinalBlock"><see langword="true"/> (default) when the input span contains the entire data to encode.
+        /// Set to <see langword="true"/> when the source buffer contains the entirety of the data to encode.
+        /// Set to <see langword="false"/> if this method is being called in a loop and if more input data may follow.
+        /// At the end of the loop, call this (potentially with an empty source buffer) passing <see langword="true"/>.</param>
         /// <returns>It returns the OperationStatus enum values:
-        /// - Done - on successful processing of the entire input span
+        /// - Done - on successful processing of the entire input span (also when <paramref name="isFinalBlock"/> is set to <see langword="false"/>)
```
Comment (GrabYourPitchforks) [inline src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: IMO the parenthetical is unnecessary and possibly confusing. The docs for _NeedMoreData_ and _InvalidData_ already call out these cases.

**Review by GrabYourPitchforks — APPROVED** [review]
Some minor nits with doc wording, but I don't feel strongly about it. Very much appreciate your improvements to the unit tests! :D

File: src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs (hunk 1 above)
Comment (gfoidl) [reply src/libraries/System.Memory/src/System/Buffers/Text/Base64Decoder.cs]: I removed the parenthetical.

File: src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs (hunk 2 above)
Comment (gfoidl) [reply src/libraries/System.Memory/src/System/Buffers/Text/Base64Encoder.cs]: I removed the parenthetical.

**Comment by danmosemsft** [comment]
@GrabYourPitchforks this is labeled breaking change. Can you please open an issue if necessary with https://github.com/dotnet/docs/issues/new?template=dotnet-breaking-change.md 

I don't [see an existing one](https://github.com/dotnet/docs/issues?q=+is%3Aissue+label%3Abreaking-change+)

**Comment by GrabYourPitchforks** [comment]
Per https://github.com/dotnet/runtime/issues/31348#issuecomment-706454725, no breaking change doc is needed.

**Pull request merged**